single-pool/js/src/transactions.ts

```diff
@@ -0,0 +1,302 @@
+    }
+
+    return transaction;
+  }
+
+  static async deposit(params: DepositParams) {
+    const { connection, pool, userWallet } = params;
+
+    if (!params.userStakeAccount == !params.depositFromDefaultAccount) {
+      throw 'must either provide userStakeAccount or true depositFromDefaultAccount';
+    }
+
+    const userStakeAccount = params.depositFromDefaultAccount
+      ? await findDefaultDepositAccountAddress(pool, userWallet)
+      : params.userStakeAccount;
+
+    const transaction = new Transaction();
+
+    const mint = findPoolMintAddress(this.programId, pool);
+    const poolStakeAuthority = findPoolStakeAuthorityAddress(this.programId, pool);
+    const userAssociatedTokenAccount = getAssociatedTokenAddressSync(mint, userWallet);
+
+    const userTokenAccount = params.userTokenAccount || userAssociatedTokenAccount;
+    const userLamportAccount = params.userLamportAccount || userWallet;
+    const userWithdrawAuthority = params.userWithdrawAuthority || userWallet;
+
+    if (
+      userTokenAccount.equals(userAssociatedTokenAccount) &&
+      (await connection.getAccountInfo(userAssociatedTokenAccount)) == null
+    ) {
+      transaction.add(
+        createAssociatedTokenAccountInstruction(
+          userWallet,
+          userAssociatedTokenAccount,
+          userWallet,
+          mint,
+        ),
+      );
+    }
+
+    transaction.add(
+      StakeProgram.authorize({
+        stakePubkey: userStakeAccount as PublicKey,
+        authorizedPubkey: userWithdrawAuthority,
+        newAuthorizedPubkey: poolStakeAuthority,
+        stakeAuthorizationType: StakeAuthorizationLayout.Staker,
+      }),
+    );
+
+    transaction.add(
+      StakeProgram.authorize({
+        stakePubkey: userStakeAccount as PublicKey,
+        authorizedPubkey: userWithdrawAuthority,
+        newAuthorizedPubkey: poolStakeAuthority,
+        stakeAuthorizationType: StakeAuthorizationLayout.Withdrawer,
+      }),
+    );
+
+    transaction.add(
+      SinglePoolInstruction.depositStake(
+        pool,
+        userStakeAccount as PublicKey,
+        userTokenAccount,
+        userLamportAccount,
+      ),
+    );
+
+    return transaction;
+  }
+
+  static async withdraw(params: WithdrawParams) {
+    const { connection, pool, userWallet, userStakeAccount, tokenAmount, createStakeAccount } =
+      params;
+
+    if (typeof tokenAmount == 'bigint' && tokenAmount > BigInt(Number.MAX_SAFE_INTEGER)) {
+      throw 'cannot convert tokenAmount to Number';
+    }
+
+    const transaction = new Transaction();
+
+    const poolMintAuthority = findPoolMintAuthorityAddress(this.programId, pool);
+
+    const userStakeAuthority = params.userStakeAuthority || userWallet;
+    const userTokenAccount =
+      params.userTokenAccount ||
+      getAssociatedTokenAddressSync(findPoolMintAddress(this.programId, pool), userWallet);
+    const userTokenAuthority = params.userTokenAuthority || userWallet;
```

Comment:
I'm surprised to see this! But I guess it makes sense since there's no signers... you could use a throwaway keypair and sign, but then they're screwed if they want to add anything else.  Was that your thinking roughly? Or am I missing something?

Reply:
sorry im not clear on whats surprising about this, or what you mean by signing with a throwaway keypair. the args here are basically "we assume a normal user will have a wallet address which they use for their token and stake account ownership, but we support people with more niche use-cases who need to specify the pubkeys explicitly." these functions all generate unsigned transactions, the caller is responsible for signing

Reply:
Oh nvm, this is even more clever than I thought. I know you're a big fan of delegates, so I was surprised to see you default an approval to the wallet itself. I thought this would look just like a normal transfer signed by the wallet then, which would defeat the purpose of the delegate.

But if you approve to yourself, you still limit the amount of damage! E.g. your account has 1000 SOL, you approve 100 to yourself, so when you call into another program, the most that it can take is 100 SOL.

We should talk up this flow to wallets more, because it's very clever!

Reply:
Per offline discussion, I misunderstood what this is doing. This is approving to the pool's authority, and not to the wallet. So this is perfect as is. Resolving!